### PR TITLE
Add Global Container Defaults to Assignment Configuration

### LINKED ISSUE #3255: Add Global Docker Configuration Fields to Config JSON
Add editable global fields for docker use in config json.

### PR DESCRIPTION
**Pull request by emaicus**
Closes #3255

Added global ```container_options``` object to the autograding configuration json. Within, default values for ```container_image```, ```use_router```, and ```single_port_per_container``` may be specified.  

Changed the global flag ```docker_enabled``` to be a new string ```autograding_method```, which can currently be specified as ```docker``` or ```jailed sandbox```. If ```autograding_method``` is not set, ```jailed sandbox``` is used by default. If an assignment built before this pull request is graded, ```jailed sandbox``` is assumed.

### CODE REVIEW

**Comment by emaicus** [comment]
This is failing because the Tutorial Repo hasn't yet been updated to use ```autograding_method```. 

This will be fixed by PR #21 in the Tutorial Repo.

File: grading/load_config_json.cpp [hunk 1]
```diff
@@ -108,41 +108,50 @@ void ArchiveValidatedFiles(nlohmann::json &whole_config) {
 
 void AddDockerConfiguration(nlohmann::json &whole_config) {
 
-  if (!whole_config["docker_enabled"].is_boolean()){
-    whole_config["docker_enabled"] = false;
+  //check if docker is globally enabled.
+  if (!whole_config["autograding_method"].is_string()) {
+    whole_config["autograding_method"] = "jailed sandbox";
```
Comment (bmcutler) [inline grading/load_config_json.cpp]: avoid spaces... switch to jailed_sandbox

File: grading/load_config_json.cpp [hunk 2]
```diff
@@ -211,7 +217,7 @@ void AddDockerConfiguration(nlohmann::json &whole_config) {
 
       if(this_testcase["containers"][container_num]["container_image"].is_null()){
         //TODO: store the default system image somewhere and fill it in here.
-        this_testcase["containers"][container_num]["container_image"] = "ubuntu:custom";
```
Comment (bmcutler) [inline grading/load_config_json.cpp]: where does ubuntu:custom come from now?

File: grading/load_config_json.cpp (hunk 1 above)
Comment (emaicus) [reply grading/load_config_json.cpp]: Fixed.

File: grading/load_config_json.cpp (hunk 2 above)
Comment (emaicus) [reply grading/load_config_json.cpp]: It comes from the assignment's default container. If there are no defaults set, the default field is populated as being ubuntu:custom on line 126.

**Pull request merged**